File: leetcode-04/solutions/cpp/0129.cpp
class Solution {
 public:
  int sumNumbers(TreeNode* root) {
    int ans = 0;
    int path = 0;

    dfs(root, path, ans);

    return ans;
  }

 private:
  void dfs(TreeNode* root, int& path, int& ans) {
    if (!root) return;
    if (!root->left && !root->right) {
      path = path * 10 + root->val;
      ans += path;
      path = (path - root->val) / 10;
      return;
    }

    path = path * 10 + root->val;
    dfs(root->left, path, ans);
    dfs(root->right, path, ans);
    path = (path - root->val) / 10;
  }
};

File: leetcode-04/solutions/python3/0129.py
class Solution:
    def sumNumbers(self, root: TreeNode) -> int:
        self.dfs(root, 0)

        return self.ans

    ans = 0

    def dfs(self, root: TreeNode, path: int) -> None:
        if not root:
            return
        if not root.left and not root.right:
            self.ans += (path * 10 + root.val)
            return

        self.dfs(root.left, path * 10 + root.val)
        self.dfs(root.right, path * 10 + root.val)
